client/app/dbaas/logs/detail/inputs/add/configure/logs-inputs-add-configure.controller.js

```diff
@@ -0,0 +1,399 @@
+class LogsInputsAddConfigureCtrl {
+    constructor ($q, $state, $stateParams, ControllerHelper, LogsInputsService, LogsInputsConfigureConstant, CloudMessage) {
+        this.$q = $q;
+        this.$state = $state;
+        this.$stateParams = $stateParams;
+        this.serviceName = this.$stateParams.serviceName;
+        this.inputId = this.$stateParams.inputId;
+        this.ControllerHelper = ControllerHelper;
+        this.LogsInputsService = LogsInputsService;
+        this.LogsInputsConfigureConstant = LogsInputsConfigureConstant;
+        this.CloudMessage = CloudMessage;
+        this.configuration = {
+            engineType: "",
+            flowgger: {},
+            logstash: {}
+        };
+        this._initLoaders();
+    }
+
+    /**
+     * initializes the input log url
+     *
+     * @memberof LogsInputsAddConfigureCtrl
+     */
+    _initLoaders () {
+        this.input = this.ControllerHelper.request.getHashLoader({
+            loaderFunction: () => this.LogsInputsService.getInput(this.serviceName, this.inputId)
+                .then(input => {
+                    this.configuration.engineType = input.info.engine.name;
+                    if (this.configuration.engineType === this.LogsInputsConfigureConstant.logstash) {
+                        this._initLogstash(input.info.engine.configuration);
+                    } else {
+                        this._initFlowgger(input.info.engine.configuration);
+                    }
+                    return input;
+                })
+        });
+
+        this.test = this.previousTest = this.ControllerHelper.request.getHashLoader({
+            loaderFunction: () => this.LogsInputsService.getTestResults(this.serviceName, this.input.data)
+        });
+
+        this.input.load()
+            .then(() => this.previousTest.load());
+    }
+
+    htmlDecode (s) {
```

Comment:
I don't really like this function, but I understand that you need it to decode a field from the API.
Anyway, this should be moved in a service. Some comments on what it does would be welcome, if the code comes from somewhere (other than sunrise), a link to the source would be appreciated as well.

Reply:
this code is coming from sunrise. Its used to decode configuration, more like URL decoding. We will move it to service and also will explore if there is an option to get rid of it.

Reply:
decodeUrlComponent() can be used. We are working on it

Reply:
Have replaced data-ng-model with ng-bind-html which has dependency on ngSanitize. This actually decodes most of html entities. I dont know if it covers all cases from htmldecode function. Have to test this to know if something is missing.